### PR TITLE
Check that we do have a port available.

### PR DESCRIPTION
**Pull request by obino**
Before we were assuming we would find a port. We now actively check if the
port is available.

### CODE REVIEW

File: AppController/djinn.rb [hunk 1]
```diff
@@ -5139,6 +5148,10 @@ def add_appserver_process(app_id, secret)
       end
   
       appengine_port = find_lowest_free_port(STARTING_APPENGINE_PORT)
+      if appengine_port < 0
+        Djinn.log_error("Failed to get a port for  #{app}")
```
Comment (menivaitsi) [inline AppController/djinn.rb]: Remove extra whitespace before #{app}.

File: AppController/lib/helperfunctions.rb [hunk 2]
```diff
@@ -1496,7 +1496,8 @@ def self.get_app_thread_safe(app)
   # Raises:
   #   SystemExit: Always occurs, since this method crashes the AppController.
   def self.log_and_crash(message)
-    self.write_file(APPCONTROLLER_CRASHLOG_LOCATION, message)
+    self.write_file(APPCONTROLLER_CRASHLOG_LOCATION, Time.new.to_s + ": "
+      +  message)
```
Comment (menivaitsi) [inline AppController/lib/helperfunctions.rb]: remove extra whitespace after +

**Comment by menivaitsi** [comment]
Is this in case we have no more ports available for deploying an app?

**Comment by obino** [comment]
OCD run https://ocd.appscale.com:8080/job/Daily%20Build/503/?auto_refresh=true

File: AppController/djinn.rb [hunk 3]
```diff
@@ -623,18 +613,28 @@ def relocate_app(appid, http_port, https_port, secret)
...
     http_port = Integer(http_port)
     https_port = Integer(https_port)
 
     # First, only let users relocate apps to ports that the firewall has open
     # for App Engine apps.
-    if http_port != 80 and (http_port < 8080 or http_port > 8100)
-      return "Error: HTTP port must be 80, or in the range 8080-8100."
+    if http_port != 80 and
+       (http_port < Nginx::START_PORT or http_port > Nginx::END_PORT)
+      return "Error: HTTP port must be 80, or in the range #{START_PORT}-#{END_PORT}."
```
Comment (cdonati) [inline AppController/djinn.rb]: Should this be #{Nginx::START_PORT}, etc?

Comment (obino) [inline AppController/djinn.rb]: yes I think so: good watch.

**Comment by obino** [comment]
@menivaitsi  yes this patch is trying to prevent silent failures if we run out or ports. We would probably need a better way to define them (for example how many ports we want to use at most in AppScalefile or something similar) since at this time it's in the code.

File: AppController/djinn.rb [hunk 4]
```diff
@@ -623,18 +613,29 @@ def relocate_app(appid, http_port, https_port, secret)
       return BAD_SECRET_MSG
     end
 
+    # Only the login node runs relocate.
+    if not my_node.is_login?
+      Djinn.log_warn("Only login nodes runs relocate")
```
Comment (menivaitsi) [inline AppController/djinn.rb]: Add a period at the end.

File: AppController/djinn.rb [hunk 5]
```diff
@@ -623,18 +613,29 @@ def relocate_app(appid, http_port, https_port, secret)
...
+      return "Error: this is not the login node"
```
Comment (menivaitsi) [inline AppController/djinn.rb]: Same here.

File: AppController/djinn.rb [hunk 6]
```diff
@@ -1359,7 +1358,13 @@ def get_deployment_id(secret)
       return BAD_SECRET_MSG
     end
 
-    return ZKInterface.get('/deployment_id')
+    begin
+      return ZKInterface.get('/deployment_id')
+    rescue FailedZooKeeperOperationException => e
+      Djinn.log_warn("(get_deployment_id) failed talking to zookeeper " +
```
Comment (menivaitsi) [inline AppController/djinn.rb]: Same for all logs here.

File: AppController/djinn.rb [hunk 7]
```diff
@@ -1417,13 +1427,16 @@ def stop_app(app_name, secret)
 
       # Contact the soap server and remove the application
       if (@app_names.include?(app_name) and !my_node.is_appengine?) or @nodes.length == 1
-        ip = HelperFunctions.read_file("#{CONFIG_FILE_LOCATION}/masters")
-        uac = UserAppClient.new(ip, @@secret)
+        uac = UserAppClient.new(@userappserver_private_ip, @@secret)
         begin
+          if not uac.does_app_exist?(app_name)
+            Djinn.log_info("(stop_app) #{app_name} does not exists")
+            return "Application #{app_name} does not exists."
```
Comment (menivaitsi) [inline AppController/djinn.rb]: exists -> exist

File: AppController/djinn.rb [hunk 8]
```diff
@@ -4748,18 +4814,25 @@ def setup_appengine_application(app, is_new_app)
   # Returns:
   #   A Fixnum corresponding to the port number that a new process can be bound
   #   to.
-  def find_lowest_free_port(starting_port)
+  def find_lowest_free_port(starting_port, ending_port=0)
     possibly_free_port = starting_port
     loop {
+      # if we have ending_port, we need to check the upper limit too.
```
Comment (menivaitsi) [inline AppController/djinn.rb]: if -> If

File: AppController/djinn.rb [hunk 9]
```diff
@@ -5511,9 +5594,9 @@ def copy_app_to_local(appname)
     return false
   end
 
+  # This function creates the xmpp account for the app for app named 'app',
```
Comment (menivaitsi) [inline AppController/djinn.rb]: "This function creates the xmpp account for the app for app named 'app'," -> 
"This function creates the xmpp account for 'app'," ?

File: AppController/lib/nginx.rb [hunk 10]
```diff
@@ -36,7 +36,11 @@ module Nginx
   MAIN_CONFIG_FILE = File.join(NGINX_PATH, "nginx.#{CONFIG_EXTENSION}")
 
 
+  # These ports are the one visible from outside, ie the ones that we
+  # attach to running application. Default is to have a maximum of 21
```
Comment (menivaitsi) [inline AppController/lib/nginx.rb]: application -> applications

File: AppController/lib/nginx.rb [hunk 11]
```diff
@@ -36,7 +36,11 @@ module Nginx
...
+  # application (8080-8100).
```
Comment (menivaitsi) [inline AppController/lib/nginx.rb]: Same here.

**Comment by obino** [comment]
OCD run https://ocd.appscale.com:8080/job/Daily%20Build/513/

**Pull request merged**